src/kv/KeyValueDB.h

```diff
@@ -226,77 +246,138 @@ class KeyValueDB {
+                     generic_iter->seek_to_last(prefix);
     }
     int upper_bound(const std::string &after) override {
-      return generic_iter->upper_bound(prefix, after);
+      return is_cf ? cf_iter->upper_bound(after) :
+                     generic_iter->upper_bound(prefix, after);
     }
     int lower_bound(const std::string &to) override {
-      return generic_iter->lower_bound(prefix, to);
+      return is_cf ? cf_iter->lower_bound(to) :
+                     generic_iter->lower_bound(prefix, to);
     }
     bool valid() override {
-      if (!generic_iter->valid())
-	return false;
-      return generic_iter->raw_key_is_prefixed(prefix);
+      if (is_cf) {
+        return cf_iter->valid();
+      } else {
+        if (!generic_iter->valid())
+	  return false;
+        return generic_iter->raw_key_is_prefixed(prefix);
+      }
     }
     // Note that next() and prev() shouldn't validate iters,
     // it's responsibility of caller to ensure they're valid.
     int next(bool validate=true) override {
       if (validate) {
         if (valid())
-          return generic_iter->next();
+          return is_cf ? cf_iter->next() : generic_iter->next();
         return status();
       } else {
-        return generic_iter->next();  
+        return is_cf ? cf_iter->next() : generic_iter->next();  
       }      
     }
     
     int prev(bool validate=true) {
       if (validate) {
         if (valid())
-          return generic_iter->prev();
+          return is_cf ? cf_iter->prev() : generic_iter->prev();
         return status();
       } else {
-        return generic_iter->prev();  
+        return is_cf ? cf_iter->prev() : generic_iter->prev();
       }      
     }
     std::string key() override {
-      return generic_iter->key();
+      return is_cf ? cf_iter->key() : generic_iter->key();
     }
+    // For column family, keys actually don't contain prefix;
+    // to make interface compatible, prefix is added.
     std::pair<std::string, std::string> raw_key() {
-      return generic_iter->raw_key();
+      return is_cf ? make_pair(prefix, cf_iter->key()) : generic_iter->raw_key();
     }
     bufferlist value() override {
-      return generic_iter->value();
+      return is_cf ? cf_iter->value() : generic_iter->value();
     }
     bufferptr value_as_ptr() {
-      return generic_iter->value_as_ptr();
+      return is_cf ? cf_iter->value_as_ptr() : generic_iter->value_as_ptr();
     }
     int status() override {
-      return generic_iter->status();
+      return is_cf ? cf_iter->status() : generic_iter->status();
     }
   };
-
   typedef ceph::shared_ptr< IteratorImpl > Iterator;
 
   WholeSpaceIterator get_iterator() {
     return _get_iterator();
   }
 
+  ColumnFamilyIterator get_cf_iterator(const std::string& cf_name) {
+    return _get_cf_iterator(cf_name);
+  }
+
+  void add_column_family(const std::string& cf_name, void *handle) {
+    cf_handles.insert(std::make_pair(cf_name, handle));
```

Comment:
is it possible to mark handle as const? If we can make the pairing const with the name. Not sure how to achieve with void pointers.

Reply:
maybe ... I will try when I modify it.